### PR TITLE
Mobile search filters

### PR DESCRIPTION
**Pull request by davidpmccormick**
## Who is this for?
People who want to use filters on smaller screens

## What is it doing for them?
Giving them a better UI to do it.

mobile_filters

__TODO__
- [x] trap focus in filters modal

### CODE REVIEW

File: common/views/components/Checkbox/Checkbox.js [hunk 1]
```diff
@@ -71,6 +71,10 @@ type CheckboxWithoutLabelProps = {|
 type CheckboxProps = {|
   ...CheckboxWithoutLabelProps,
   text: string,
+  checked: ?boolean,
```
Comment (jamesgorrie) [inline common/views/components/Checkbox/Checkbox.js]: When is this ever `null`?

File: common/views/components/Checkbox/Checkbox.js [hunk 2]
```diff
@@ -71,6 +71,10 @@ type CheckboxWithoutLabelProps = {|
...
+  name: ?string,
```
Comment (jamesgorrie) [inline common/views/components/Checkbox/Checkbox.js]: `nullable`?

File: common/hooks/useFocusTrap.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+// @flow
+
+import { useEffect, type ElementRef } from 'react';
+
+type TrapElRef = {|
+  current: ElementRef<any>,
+|};
+
+export default (startRef: ?TrapElRef, endRef: ?TrapElRef) => {
```
Comment (jamesgorrie) [inline common/hooks/useFocusTrap.js]: Should we follow the components pattern, which I think helps with debugging as the function name can be spat out when sourcemaps work of

```JS
function useFocusTrap() {}
export default useFocusTrap;
```

File: common/hooks/useFocusTrap.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  function handleTrapStartKeyDown(event) {
+    if (event.shiftKey && event.keyCode === 9) {
```
Comment (jamesgorrie) [inline common/hooks/useFocusTrap.js]: Is [this](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key) better?

File: common/views/components/Checkbox/Checkbox.js (hunk 1 above)
Comment (davidpmccormick) [reply common/views/components/Checkbox/Checkbox.js]: nowhere! Will fix

File: common/views/components/Checkbox/Checkbox.js (hunk 2 above)
Comment (davidpmccormick) [reply common/views/components/Checkbox/Checkbox.js]: Nope. Fixing.

File: common/hooks/useFocusTrap.js (hunk 3 above)
Comment (davidpmccormick) [reply common/hooks/useFocusTrap.js]: yeah, good shout. Will update.

File: common/hooks/useFocusTrap.js (hunk 4 above)
Comment (davidpmccormick) [reply common/hooks/useFocusTrap.js]: Yes, almost certainly. In my head it wasn't as well supported as `keyCode` (inspite of `keyCode` being deprecated). But looks like I'm wrong about that. Will update.

File: common/views/components/Checkbox/Checkbox.js [hunk 5]
```diff
@@ -71,6 +71,10 @@ type CheckboxWithoutLabelProps = {|
 type CheckboxProps = {|
   ...CheckboxWithoutLabelProps,
   text: string,
+  checked: boolean,
+  name: string,
+  onChange: ?() => void,
```
Comment (jamesgorrie) [inline common/views/components/Checkbox/Checkbox.js]: If we send a value, doesn't react need you to send an `onChange`?

File: common/views/components/SearchFilters/SearchFilters.js [hunk 6]
```diff
@@ -46,12 +13,25 @@ type Props = {|
...
+  workTypeInUrlArray: string[],
+|};
+
 const SearchFilters = ({
   searchForm,
   searchParams,
   workTypeAggregations,
   changeHandler,
 }: Props) => {
+  const [isMobile, setIsMobile] = useState(false);
```
Comment (jamesgorrie) [inline common/views/components/SearchFilters/SearchFilters.js]: It'd be good to make this a hook too?

Similar to what we should do with `const isEnhanced = `useIsEnhanced()`

File: common/views/components/SearchFilters/SearchFilters.js [hunk 7]
```diff
@@ -107,189 +99,29 @@ const SearchFilters = ({
...
-                </a>
-              </NextLink>
-            </Space>
-          </div>
-        )}
-      </Space>
-    </div>
+  return (
+    <>
+      {isMobile ? (
```
Comment (jamesgorrie) [inline common/views/components/SearchFilters/SearchFilters.js]: Hmm. Why JS and CSS like everything else for responsive layouts?

Comment (jamesgorrie) [reply common/views/components/SearchFilters/SearchFilters.js]: We could create some helpers and just apply those classes on wrappers around the components.

File: common/views/components/SearchFilters/SearchFilters.js (hunk 6 above)
Comment (jamesgorrie) [reply common/views/components/SearchFilters/SearchFilters.js]: Actually scrap that - I thought this was going to be used for functionality - but it's used for styling.

[This?](https://github.com/wellcometrust/wellcomecollection.org/pull/4948/files#r360421941)

**Comment by jamesgorrie** [comment]
This creates quite a lot of code to maintain - where was the issue raised with the current mobile UI?

File: common/views/components/SearchFilters/SearchFilters.js (hunk 7 above)
Comment (davidpmccormick) [reply common/views/components/SearchFilters/SearchFilters.js]: Having the named form elements duplicated on the page was problematic when it came to submitting the form (and the UI diverges too much between desktop and mobile to realistically be covered off by the same elements with CSS alone).

After chatting with @gestchild, reasoned swapping them out in this manner made the most sense.

**Comment by davidpmccormick** [comment]
#4926

The current mobile UI barely scales to the two filter types (date and format) we have, and certainly won't handle a third. Unsightly horizontal scrollbars are already a thing on my small screen with the current setup.

File: common/views/components/Checkbox/Checkbox.js (hunk 5 above)
Comment (davidpmccormick) [reply common/views/components/Checkbox/Checkbox.js]: We do. Good point. Will update.

**Comment by kenoir** [comment]
What does this need to merge?

**Comment by jamesgorrie** [comment]
Not a massive fan of the fact that we now have multiple ways and methods of dealing with responsie layouts.

Could we keep the current method of wrapping things in containers that show and hide via CSS (and thus follow all other layout logic)?

e.g.
```JS
// from is inclusive, to is exclusive
<div className={display({ from: 's', to: `m`  })}></div>
<div className={display({ from: `m`, to: `xl` })}></div>
```

We've not made the decision anywhere to have the UI respond via JS, unless there is a piece of functionality that requires JS, which this doesn't.

It feels a shame to start fragmenting from a core understanding of responsive design.

**Comment by davidpmccormick** [comment]
I don't think this is 'degrading' to mobile, but rather a different, more appropriate, mobile layout.

Even if we show/hide the two layouts, rather than conditionally rendering them with JS, we'd still have to handle the fact the form elements are duplicated on the page, which creates bugs when submitting. We'd have to e.g. add/remove form element attributes – and at that point I don't know what's gained from using a CSS approach, because it would still be reliant on JS.

Having said that, I think I take the point about reflow.

**Comment by jamesgorrie** [comment]
> I don't think this is 'degrading' to mobile, but rather a different, more appropriate, mobile layout.

Do we think mobile first is still appropriate? Thinking of writing up some dev processes, and wondering if this should be included?

I always thought the point of mobile first was to start with a simple design that is likely to scale as it works on the lowest, most frugal space and availability of functionality. What is feels like is that we have designed something that doesn't scale, made it the default, and are now pathing it for mobile. It is a process that stung us on the editorial content, so just trying to mitigate repating that.

Sorry, I didn miss the point on replicated form elements - that's a tricky one...

**Comment by davidpmccormick** [comment]
I think _mobile first by default_ makes sense. What we started with (iterating from the two prototypes)  _was_ mobile first in its strictest sense, with the mobile and desktop layouts using the same elements with just some CSS amends.

But after design review and iteration, the view was taken that we shouldn't have to be slaves to the strictest of responsive layouts where a more appropriate layout was available (so long as it wasn't at the expense of content availability/comprehension).

FWIW it feels like this overarching design decision has been open-eyed, rather than a post-rationalised shoe-horning. But I agree we should be eager to sweat the implementation specifics.

**Comment by jennpb** [comment]
Just wanted to capture some of the thinking behind the two filter UIs for desktop and mobile - here you go: https://wellcome.slack.com/archives/C294K7D5M/p1578568546011400

I would like to get this deployed today and then we can revisit and refactor later on if necessary.

**Pull request merged**